xwiki-platform-core/xwiki-platform-messagestream/xwiki-platform-messagestream-api/src/main/java/org/xwiki/messagestream/internal/GroupMessageStreamNotificationFilter.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.xwiki.messagestream.internal;
+
+import java.util.Collection;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Provider;
+import javax.inject.Singleton;
+
+import org.xwiki.component.annotation.Component;
+import org.xwiki.eventstream.Event;
+import org.xwiki.messagestream.GroupMessageDescriptor;
+import org.xwiki.model.reference.DocumentReference;
+import org.xwiki.notifications.NotificationFormat;
+import org.xwiki.notifications.filters.NotificationFilterPreference;
+
+import com.xpn.xwiki.XWikiContext;
+
+/**
+ * Filter that make sure a group message (to a group) from the message stream is visible only to members of this group.
+ *
+ * @version $Id$
+ * @since 12.9RC1
+ */
+@Component
+@Singleton
+@Named("GroupMessageStreamNotificationFilter")
+public class GroupMessageStreamNotificationFilter extends AbstractMessageStreamNotificationFilter
+{
+    @Inject
+    private Provider<XWikiContext> xcontextProvider;
+
+    @Override
+    public FilterPolicy filterEvent(Event event, DocumentReference user,
+        Collection<NotificationFilterPreference> filterPreferences, NotificationFormat format)
+    {
+        FilterPolicy ret;
+        // Don't handle events that are not group messages.
+        if (!getEventType().equals(event.getType())) {
+            ret = FilterPolicy.NO_EFFECT;
+        } else if (user != null) {
+            XWikiContext xWikiContext = this.xcontextProvider.get();
+            // If the user is not in the messaged group, the event is filtered.
+            // If the user is in the group, we let it pass and let downstream filter analyze the event.
+            boolean userInGroup = xWikiContext.getWiki().getUser(user, xWikiContext).isUserInGroup(event.getStream());
```

Comment:
Sounds potentially dangerous: getUser can return null AFAIR. And don't we have a better API from GroupManager or something like this to check if a user belongs to a group?

Reply:
I'll add a null check on the result of get user.
Regarding `isUserInGroup` I didn't found anything better.

Reply:
> Regarding isUserInGroup I didn't found anything better.

interesting, IMO this is a standard API that we'd need either in User API or in Group API.

Reply:
I agree about some ``isUserInGroup`` "should be API". However I am sure it is not; there are several extensions like the [Publication Workflow](https://github.com/xwiki-contrib/workflow-publication/blob/5c975c988d0ae86da2a67299810e008ca8c74baa/xwiki-workflow-publication-api/src/main/java/org/xwiki/workflowpublication/internal/DefaultPublicationRoles.java#L98) who do a manual check (recursing over possible sub groups etc.) to implement this feature. @manuelleduc 's solution is rather short in comparison :).